proposals/1141_audit-logging/README.md

```diff
@@ -0,0 +1,323 @@
+- Maybe: Newly authenticated user is associated with “admin” RBAC. Note that the Supervisor is not directly aware of
+  RBAC, so determining this would require otherwise unnecessary calls to the Kubernetes API server, which would degrade
+  the performance of the Supervisor. It's also not clear what would constitute "admin" level access, since RBAC is
+  configurable at a very fine-grained level. On the other hand, the Supervisor is directly aware of the user's group
+  memberships, which could be logged.
+
+The Concierge's audit logs would include events such as:
+
+- Token credential request (succeeded, failed, maybe maps to admin RBAC). While already captured by the API server audit
+  logs, those should likely be set to metadata. Duplicating the event allows for more controlled capture & management of
+  data.
+- WhoAmI Request. While already captured by the API server audit logs, duplicating the event allows for more controlled
+  capture & management of data.
+
+Other events may be useful to auditors and may be included in the audit logs, such as:
+
+- Application startup with version information
+- Graceful application shutdown
+
+##### Audit Logs as Separate Log Files
+
+The Concierge and Supervisor apps could each send audit logs to separate files on disk in JSON format. The performance
+impact of logging to a file should be acceptable thanks to file buffering, but this assumption should be tested. Note
+that this approach would not guarantee that the log statement is flushed to the file before the action is performed,
+because then we would lose the benefit of buffering. It would be "best effort" to the file, e.g. the process crashing
+might lose a few lines of logs. A normal pod shutdown should be able to flush the file without any loss.
+
+[A new streaming sidecar container](https://kubernetes.io/docs/concepts/cluster-administration/logging/#sidecar-container-with-logging-agent)
+will be added to both the Concierge and Supervisor apps Deployments' Pods. These containers will tail those audit logs
+to stdout, thus effectively moving those log lines from files on the Pod to Kubernetes container logs. Those sidecar
+container images can be minimal with just enough in the image to support the unix `tail` command (or similar Go binary,
+such as [hpcloud/tail](https://github.com/hpcloud/tail)).
+
+Kubernetes will take care of concerns such as log rotation for the container logs. For the files on the Pod's disk
+output by the Supervisor and Concierge apps, we should research whether Pinniped should have code to avoid allowing
+those files from growing too large. Old lines can be discarded since the sidecar container should have already streamed
+them.
+
+Container logs in JSON format are easy for node-level logging agents, e.g. fluentbit, to ingest/annotate/parse/filter
+and send to numerous sink destinations. These containers could still run when audit logs are disabled by the admin, but
+would produce no log lines in that case.
```

Comment:
Why keep them running if they don't log anything?

Reply:
Mainly because it would make the implementation simpler. However, it might be possible to use ytt directives to avoid adding the container to the Pod when the user uses the ytt value to disable audit logging, if that is desired.